### PR TITLE
`Development`: Refine path comparison

### PR DESCRIPTION
**Pull request by MarkusPaulsen**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I documented the Java code using JavaDoc style.

### Motivation and Context
- Some file paths may address areas outside of the repository (invalid file paths).
- Naming file path parameters "filename" may spread the expectation, that this is just a file name and not a file path (with less things required to be considered).

### Description
The file paths now compared in an unescaped, normalised and absolute manner against the repository path in order to check whether the given path is valid. Also filename parameters were renamed to filePath in order to not spread incorrect expectations. 

### Steps for Testing
We just want to make sure that I didn't accidentally break the implementation.

Prerequisites:
- 1 Instructor
- 1 Student

After each step, commit and see if the changes were persisted by reloading
Programming Exercise (Instructor, Student):
1. Create a programming exercise with the online editor activated.
2. Participate in the programming exercise/alternatively edit the solution/template in the online editor
3. Create a folder
4. Create a folder within the folder
5. Create a file within the second folder
6. Rename the file
7. Delete the file and folders
8. No issues should occur

#### Performance Review
- [x] I confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| src/main/java/de/tum/in/www1/artemis/service/RepositoryService.java | 91.5% | ✅                           |

### CODE REVIEW

**Review by MaximilianAnzinger — APPROVED** [review]

**Comment by pal03377** [comment]
I wanted to test this PR, but I now get a build error for the docker image :/ 

```
#17 103.3 /opt/artemis/src/main/java/de/tum/in/www1/artemis/service/RepositoryService.java:266: error: cannot find symbol
#17 103.4         String newFilename = fileService.sanitizeFilename(fileMove.newFilename());
#17 103.4                                         ^
#17 103.4   symbol:   method sanitizeFilename(String)
#17 103.4   location: variable fileService of type FileService
```

**Comment by MaximilianAnzinger** [comment]
Is it possible that you have changes in `FileService.java` that are not committed yet? Seems like the method `FileService.sanitizeFilename(String)` is missing.

**Review by MaximilianAnzinger — APPROVED** [review]
Build error fixed 👍

**Review by MaximilianAnzinger — APPROVED** [review]
Tested on TS3. Works according to testing steps.
Code LGTM

**Review by JohannesStoehr — APPROVED** [review]
Creation and deletion is working fine for me.
When trying to delete a nested folder only the innermost folder is actually deleted, is that intended?

**Review by tobias-lippert — APPROVED** [review]
Tested in testing session and no issues occurred.

**Review by laadvo — APPROVED** [review]
Tested during the testing session on ts3, working as expected :+1:

**Comment by julian-christl** [comment]
> Creation and deletion is working fine for me. When trying to delete a nested folder only the innermost folder is actually deleted, is that intended?

@JohannesStoehr It's the current behaviour. Changing this would not be within the scope of the PR.

**Review by nityanandaz — APPROVED** [review]
Tested on TS3 during testing session (2023-08-15), could rename/create multiple levels of folders

File: src/main/java/de/tum/in/www1/artemis/service/RepositoryService.java [hunk 1]
```diff
@@ -178,47 +179,76 @@ public Map<String, Boolean> getFilesWithInformationAboutChange(Repository reposi
...
+        Path unknownInputPath = Paths.get(unescapedPath).normalize();
+        Path absoluteRepositoryPath = repository.getLocalPath().normalize().toAbsolutePath();
+        Path absoluteInputPath = absoluteRepositoryPath.resolve(unknownInputPath).normalize();
+        if (!absoluteInputPath.startsWith(absoluteRepositoryPath)) {
+            throw new IllegalArgumentException("Path is not valid");
+        }
+        return unknownInputPath;
+    }
+
+    private Path checkIfPathIsValidAndExistsAndReturnSafePath(Repository repository, String path, boolean exists) {
```
Comment (laadvo) [inline src/main/java/de/tum/in/www1/artemis/service/RepositoryService.java:224]: I'm not sure the method name is correct here. I don't think the method checks if the path exists, rather it is specified in the parameter if it supposed to exist. Or am I mistaken?

**Review by laadvo — COMMENTED** [review]
Code looks good, but you could consider adding Javadoc to `checkIfPathIsValidAndReturnSafePath` and especially `checkIfPathIsValidAndExistsAndReturnSafePath`. All other methods in the class seem to have it.

**Comment by julian-christl** [comment]
Memo: 5 Testing approvals and one Code Review approvals at this point

**Comment by julian-christl** [comment]
Memo 2: Only Method name and JavaDoc changed. No retest required

**Review by laadvo — APPROVED** [review]
Thanks for considering my suggestion, code looks good :+1:

**Review by tobias-lippert — APPROVED** [review]
Reapprove

**Review by nityanandaz — APPROVED** [review]

**Review by JohannesStoehr — APPROVED** [review]
Reapprove

**Pull request merged**